web/impact/impact/v1/views/office_hours_calendar_view.py

```diff
@@ -327,22 +337,44 @@ def _make_f_expression(field, location_path):
     return F("{}__{}".format(location_path, field))
 
 
-def _date_range(date_spec=None):
+def _date_range(calendar_span, upcoming, focal_date=None):
     # returns (start_date, end_date)
+    # When calendar_span == week and upcoming == False
     # start_date is the latest monday that is less than or equal to today,
     # end_date is start_date + seven days
+    # When calendar_span == month and upcoming == False
+    # start_date is the first day of the new month
+    # end_date is start_date + length of the current month
     # both values are then padded by 24 hours to allow for TZ differences
+    # when upcoming == True
+    # start_date is the current date period
 
-    # throws ValueError if date_spec is not in ISO-8601 format
+    # throws ValueError if focal_date is not in ISO-8601 format
 
-    if date_spec:
-        initial_date = datetime.strptime(date_spec, ISO_8601_DATE_FORMAT)
+    if focal_date:
+        initial_date = datetime.strptime(focal_date, ISO_8601_DATE_FORMAT)
     else:
         initial_date = datetime.now()
+
+    if upcoming and initial_date < datetime.now():
```

Comment:
The `upcoming` parameter, provided by the client, seems to exist only to override the date provided by the frontend client. This seems confusing and unnecessary. 
My understanding is that we providing the office hour sessions for a given calendar month. If so, we should do that. 
If not, maybe we need to clarify the objective.

Reply:
We want to provide office hour for a calendar month, however sometimes we filter out out those that are past the current date, as as user shouldn't be able to reserve past office hours

Reply:
@jonkiparsky Unlike the new office hours on accelerating, the directory is only interested in upcoming office hours since it's only meant to be used by finalists for reservation.

Reply:
Okay, so it seems like the simplest thing to do is to let the front end deal with this - either display past sessions as unreservable, or just drop them on the floor.

Doing this in the back end adds needless complexity both to the function itself and to the interface that it defines.

Reply:
The front end code for this part of the interface is already involved, I think its best the back-end handle this filtering; adding another layer of complexity whereby the interface will loop through the list of return office hours will further complicate it.
Also if the backend handle this we get uniform performance across devices. and we don't have to transport as much data through the wire.
@jonkiparsky  @geofrocker your thought please

Reply:
- The amount of data is a red herring, IMO. I think it's unlikely that a user will create enough office hours in a month for this to make a difference, and in any case, if the amount of data in a full month's sessions is problematic then we're going to have that problem at the start of month, so we'll have to deal with it regardless. 
- I would assume that the front end is already looping over the data in order to display it. This would simply ask it to check if the date whether the session is in the past in that already-existing loop.
- If we want to do this in the back end, we'll have to go back to the drawing board and redesign the endpoint's interface and also refactor the implementation. Having a parameter that nullifies other parameters will certainly lead to problems.

Reply:
You make a good points here: 

- the frontend does not loop over the sessions, at least our code doesn't however the calendar package we use does but outside of our codebase.

- Given that the other option is to re-write the endpoint we can accept the trade off and do the implementation on the front end.